identity/service_principals.go

```diff
@@ -0,0 +1,300 @@
+func (a ServicePrincipalsAPI) VerifyServicePrincipalAsAdmin(servicePrincipalID string, adminGroupID string) (bool, error) {
+	servicePrincipal, err := a.read(servicePrincipalID)
+	if err != nil {
+		return false, err
+	}
+	for _, group := range servicePrincipal.Groups {
+		if group.Value == adminGroupID {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// RemoveServicePrincipalAsAdmin will remove the servicePrincipal from the admin group given the admin group id and servicePrincipal id
+func (a ServicePrincipalsAPI) RemoveServicePrincipalAsAdmin(servicePrincipalID string, adminGroupID string) error {
+	servicePrincipalPath := fmt.Sprintf("/preview/scim/v2/ServicePrincipals/%v", servicePrincipalID)
+	var removeOperations ServicePrincipalPatchOperations
+	servicePrincipalPatchRequest := ServicePrincipalPatchRequest{
+		Schemas:    []URN{PatchOp},
+		Operations: []ServicePrincipalPatchOperations{},
+	}
+	path := fmt.Sprintf("groups[value eq \"%s\"]", adminGroupID)
+	removeOperations = ServicePrincipalPatchOperations{
+		Op:   "remove",
+		Path: path,
+	}
+	servicePrincipalPatchRequest.Operations = append(servicePrincipalPatchRequest.Operations, removeOperations)
+	return a.C.Scim(http.MethodPatch, servicePrincipalPath, servicePrincipalPatchRequest, nil)
+}
+
+// GetOrCreateDefaultMetaServicePrincipal ...
+func (a ServicePrincipalsAPI) GetOrCreateDefaultMetaServicePrincipal(metaServicePrincipalDisplayName string, metaServicePrincipalName string, deleteAfterCreate bool) (servicePrincipal ScimServicePrincipal, err error) {
+	var servicePrincipals ServicePrincipalList
+	err = a.C.Scim(http.MethodGet, "/preview/scim/v2/ServicePrincipals", map[string]string{
+		"filter": "displayName+eq+" + metaServicePrincipalDisplayName,
+	}, &servicePrincipals)
+	if err != nil {
+		return servicePrincipal, err
+	}
+	resources := servicePrincipals.Resources
+	if len(resources) == 1 {
+		return resources[0], err
+	} else if len(resources) > 1 {
+		return ScimServicePrincipal{}, errors.New("more than one meta servicePrincipal")
+	}
+
+	log.Printf("Meta ServicePrincipal not found will create a new meta servicePrincipal with name: %s\n", metaServicePrincipalDisplayName)
+
+	newCreatedServicePrincipal, err := a.Create(metaServicePrincipalName, metaServicePrincipalDisplayName, nil, nil)
+	if err != nil {
+		if strings.Contains(err.Error(), "already exists") {
+			time.Sleep(time.Second * 1)
+			return a.GetOrCreateDefaultMetaServicePrincipal(metaServicePrincipalDisplayName, metaServicePrincipalName, deleteAfterCreate)
+		}
+		return servicePrincipal, err
+	}
+	if deleteAfterCreate {
+		defer func() {
+			deferErr := a.Delete(newCreatedServicePrincipal.ID)
+			err = deferErr
+		}()
+	}
+	return newCreatedServicePrincipal, err
+}
+
+func (a ServicePrincipalsAPI) getInheritedAndNonInheritedRoles(servicePrincipal ScimServicePrincipal, groups []ScimGroup) (inherited []RoleListItem, unInherited []RoleListItem) {
```

Comment:
remove this one.

Reply:
2c313ea0d30a96f5ec12d5aa1572c2fc02c32bd9